src/Elastic.CommonSchema/Serialization/MetadataDictionaryConverter.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Elastic.CommonSchema.Serialization
+{
+	internal class MetadataDictionaryConverter : JsonConverter<MetadataDictionary>
+	{
+		public override MetadataDictionary Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+		{
+			if (reader.TokenType != JsonTokenType.StartObject)
+				throw new JsonException($"JsonTokenType was of type {reader.TokenType}, only objects are supported");
+
+			var dictionary = new MetadataDictionary();
+			while (reader.Read())
+			{
+				if (reader.TokenType == JsonTokenType.EndObject)
+					return dictionary;
+
+				if (reader.TokenType != JsonTokenType.PropertyName)
+					throw new JsonException("JsonTokenType was not PropertyName");
+
+				var propertyName = reader.GetString();
+
+				if (string.IsNullOrWhiteSpace(propertyName))
+					throw new JsonException("Failed to get property name");
+
+				reader.Read();
+
+				dictionary.Add(propertyName, ExtractValue(ref reader, options));
+			}
+
+			return dictionary.Count > 0 ? dictionary : null;
+		}
+
+		public override void Write(Utf8JsonWriter writer, MetadataDictionary value, JsonSerializerOptions options)
+		{
+			writer.WriteStartObject();
+
+			foreach (var kvp in value)
+			{
+				var propertyName = kvp.Key;
+				writer.WritePropertyName(propertyName);
+
+				if (kvp.Value == null)
+					writer.WriteNullValue();
+				else
+				{
+					var inputType = kvp.Value.GetType();
+					//TODO prevent reentry and cache get converters
```

Comment:
Is this intended for a future PR?

Reply:
No that's a stale TODO in older versions of System.Text.Json it was worthwhile performance wise to call directly into any known converter:

https://github.com/elastic/ecs-dotnet/pull/211/files/229ac063e5981c3305a6bb72f11a49e1be73ca94#diff-ed861ef4bf6340af5d35ed6493d67a8f071fdd6cb2c80d125cbc4333499d3f5eL73-L74

However this no longer works nor has any benefit over calling `.Deserialize<>()`.